jwst/outlier_detection/utils.py

```diff
@@ -20,177 +26,365 @@
 _ONE_MB = 1 << 20
 
 
+def nanmedian3D(cube, overwrite_input=True):
+    """Compute the median of a cube ignoring warnings and with memory efficiency.
+    np.nanmedian always uses at least 64-bit precision internally, and this is too
+    memory-intensive. Instead, loop over the median calculation to avoid the
+    memory usage of the internal upcasting and temporary array allocation.
+    The additional runtime of this loop is indistinguishable from zero,
+    but this loop cuts overall step memory usage roughly in half for at least one
+    test association.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings(action="ignore",
+                                message="All-NaN slice encountered",
+                                category=RuntimeWarning)
+        output_arr = np.empty(cube.shape[1:], dtype=np.float32)
+        for i in range(output_arr.shape[0]):
```

Comment:
I think this is a new spot where we are picking a "buffer size". If i'm understanding this takes a nanmedian across axis 0 for all of axis 2 for each collection (row/col) in axis 1. So for a median cube of [30, 2048, 1024] this loop will run 2048 times. As this cube is [n_groups, output_image_shape[0], output_image_shape[1]] do you have a sense for how this will scale for large fields of view?

Also, does this loop trade runtime for memory usage (longer runtime -> less memory used)? Any comments on what impact this has?

Reply:
Thanks for the details. That all makes sense. It seems like the axis along which runtime should diverge most (from a `np.nanmedian`) would be when axis1 grows (causing more iterations through the for loop). I did some local testing of this with various shapes including a few from the table Hien shared (good idea!). Things were mostly as expected and as you described. However one pleasant surprise was a reduction in runtime for large cube. Testing with 2 cube: (6, 40162, 19473) and (3, 40542, 20091) (taken from the spreadsheet but with the first axis reduced to allow them to run on my machine). Runtimes for an equivalent `np.nanmedian` calls were 208 and 128 seconds. However `nanmedian3D` only took 85 and 55 seconds.